File: typescript/src/configuration/kernelConfig.ts
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { AzureOpenAIConfig, OpenAIConfig } from '../ai';
import { Verify } from '../diagnostics';
import { IRetryMechanism, PassThroughWithoutRetry } from '../reliability';
import { BackendTypes, IBackendConfig } from './iBackendConfig';

export class KernelConfig {
    private _completionBackends: { [key: string]: IBackendConfig } = {};
    private _embeddingsBackends: { [key: string]: IBackendConfig } = {};
    private _defaultCompletionBackend?: string;
    private _defaultEmbeddingsBackend?: string;
    private _retryMechanism: IRetryMechanism = new PassThroughWithoutRetry();

    /**
     * Global retry logic used for all the backends.
     */
    public get retryMechanism(): IRetryMechanism {
        return this._retryMechanism;
    }

    /**
     * Adds an Azure OpenAI backend to the list.
     * See https://learn.microsoft.com/azure/cognitive-services/openai for service details.
     *
     * @param {string} label An identifier used to map semantic functions to backend,
     * decoupling prompts configurations from the actual model used
     * @param {string} deploymentName Azure OpenAI deployment name, see https://learn.microsoft.com/azure/cognitive-services/openai/how-to/create-resource
     * @param {string} endpoint Azure OpenAI deployment URL, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart
     * @param {string} apiKey Azure OpenAI API key, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart
     * @param {string} apiVersion Azure OpenAI API version, see https://learn.microsoft.com/azure/cognitive-services/openai/reference
     * @param {boolean} overwrite Whether to overwrite an existing configuration if the same label exists
     * @returns {Self} Self instance
     */
    public addAzureOpenAICompletionBackend(
        label: string,
        deploymentName: string,
        endpoint: string,
        apiKey: string,
        apiVersion: string = '2022-12-01',
        overwrite: boolean = false
    ): KernelConfig {
        if (!label) {
            throw new Error('The label is empty');
        }

        if (!overwrite && this._completionBackends[label]) {
            throw new Error(`The completion backend cannot be added twice: ${label}`);
        }

        this._completionBackends[label] = {
            backendType: BackendTypes.AzureOpenAI,
            azureOpenAI: AzureOpenAIConfig.create(deploymentName, endpoint, apiKey, apiVersion),
        };

        if (Object.keys(this._completionBackends).length === 1) {
            this._defaultCompletionBackend = label;
        }

        return this;
    }

    /**
     * Adds the OpenAI completion backend to the list.
     * See https://platform.openai.com/docs for service details.
     *
     * @param {string} label An identifier used to map semantic functions to backend,
     * decoupling prompts configurations from the actual model used
     * @param {string} modelId OpenAI model name, see https://platform.openai.com/docs/models
     * @param {string} apiKey OpenAI API key, see https://platform.openai.com/account/api-keys
     * @param {string} orgId OpenAI organization id. This is usually optional unless your account belongs to multiple organizations.
     * @param {boolean} overwrite Whether to overwrite an existing configuration if the same label exists
     * @returns {this} Self instance
     */
    public addOpenAICompletionBackend(
        label: string,
        modelId: string,
        apiKey: string,
        orgId?: string,
        overwrite: boolean = false
    ): KernelConfig {
        if (!label) {
            throw new Error('The label is empty');
        }

        if (!overwrite && this._completionBackends[label]) {
            throw new Error(`The completion backend cannot be added twice: ${label}`);
        }

        this._completionBackends[label] = {
            backendType: BackendTypes.OpenAI,
            openAI: OpenAIConfig.create(modelId, apiKey, orgId),
        };

        if (Object.keys(this._completionBackends).length === 1) {
            this._defaultCompletionBackend = label;
        }

        return this;
    }

    /**
     * Adds an Azure OpenAI embeddings backend to the list.
     * See https://learn.microsoft.com/azure/cognitive-services/openai for service details.
     *
     * @param {string} label An identifier used to map semantic functions to backend,
     * decoupling prompts configurations from the actual model used
     * @param {string} deploymentName Azure OpenAI deployment name, see https://learn.microsoft.com/azure/cognitive-services/openai/how-to/create-resource
     * @param {string} endpoint Azure OpenAI deployment URL, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart
     * @param {string} apiKey Azure OpenAI API key, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart
     * @param {string} [apiVersion="2022-12-01"] Azure OpenAI API version, see https://learn.microsoft.com/azure/cognitive-services/openai/reference
     * @param {boolean} [overwrite=false] Whether to overwrite an existing configuration if the same label exists
     * @returns {KernelConfig} Self instance
     */
    public addAzureOpenAIEmbeddingsBackend(
        label: string,
        deploymentName: string,
        endpoint: string,
        apiKey: string,
        apiVersion: string = '2022-12-01',
        overwrite: boolean = false
    ): KernelConfig {
        Verify.notEmpty(label, 'The label is empty');

        if (!overwrite && this._embeddingsBackends[label]) {
            throw new Error(`The embeddings backend cannot be added twice: ${label}`);
        }

        this._embeddingsBackends[label] = {
            backendType: BackendTypes.AzureOpenAI,
            azureOpenAI: AzureOpenAIConfig.create(deploymentName, endpoint, apiKey, apiVersion),
        };

        if (Object.keys(this._embeddingsBackends).length === 1) {
            this._defaultEmbeddingsBackend = label;
        }

        return this;
    }

    /**
     * Adds the OpenAI embeddings backend to the list.
     * See https://platform.openai.com/docs for service details.
     *
     * @param {string} label An identifier used to map semantic functions to backend,
     * decoupling prompts configurations from the actual model used
     * @param {string} modelId OpenAI model name, see https://platform.openai.com/docs/models
     * @param {string} apiKey OpenAI API key, see https://platform.openai.com/account/api-keys
     * @param {string} [orgId] OpenAI organization id. This is usually optional unless your account belongs to multiple organizations.
     * @param {boolean} [overwrite=false] Whether to overwrite an existing configuration if the same label exists
     * @returns {KernelConfig} Self instance
     */
    public addOpenAIEmbeddingsBackend(
        label: string,
        modelId: string,
        apiKey: string,
        orgId?: string,
        overwrite: boolean = false
    ): KernelConfig {
        Verify.notEmpty(label, 'The label is empty');

        if (!overwrite && this._embeddingsBackends[label]) {
            throw new Error(`The embeddings backend cannot be added twice: ${label}`);
        }

        this._embeddingsBackends[label] = {
            backendType: BackendTypes.OpenAI,
            openAI: OpenAIConfig.create(modelId, apiKey, orgId),
        };

        if (Object.keys(this._embeddingsBackends).length === 1) {
            this._defaultEmbeddingsBackend = label;
        }

        return this;
    }
    public hasCompletionBackend(label: string, condition?: (backendConfig: IBackendConfig) => boolean): boolean {
        return condition === undefined
            ? this._completionBackends.hasOwnProperty(label)
            : Object.keys(this._completionBackends).some(
                  (key) => key === label && condition(this._completionBackends[key])
              );
    }

    public hasEmbeddingsBackend(label: string, condition?: (backendConfig: IBackendConfig) => boolean): boolean {
        return condition === undefined
            ? this._embeddingsBackends.hasOwnProperty(label)
            : Object.keys(this._embeddingsBackends).some(
                  (key) => key === label && condition(this._embeddingsBackends[key])
              );
    }

    public setRetryMechanism(retryMechanism?: IRetryMechanism): this {
        this._retryMechanism = retryMechanism ?? new PassThroughWithoutRetry();
        return this;
    }

    public setDefaultCompletionBackend(label: string): this {
        if (!this._completionBackends.hasOwnProperty(label)) {
            throw new Error(`The completion backend doesn't exist: ${label}`);
        }

        this._defaultCompletionBackend = label;
        return this;
    }

    public get defaultCompletionBackend(): string | undefined {
        return this._defaultCompletionBackend;
    }

    public setDefaultEmbeddingsBackend(label: string): KernelConfig {
        if (!this._embeddingsBackends.hasOwnProperty(label)) {
            throw new Error(`The embeddings backend doesn't exist: ${label}`);
        }

        this._defaultEmbeddingsBackend = label;
        return this;
    }

    public get defaultEmbeddingsBackend(): string | undefined {
        return this._defaultEmbeddingsBackend;
    }

    public getCompletionBackend(name?: string): IBackendConfig {
        if (name === undefined || name === '') {
            if (this._defaultCompletionBackend === undefined) {
                throw new Error(`Completion backend not found: ${name}. No default backend available.`);
            }

            return this._completionBackends[this._defaultCompletionBackend];
        }

        if (this._completionBackends.hasOwnProperty(name)) {
            return this._completionBackends[name];
        }

        if (this._defaultCompletionBackend !== undefined) {
            return this._completionBackends[this._defaultCompletionBackend];
        }

        throw new Error(`Completion backend not found: ${name}. No default backend available.`);
    }

    public getEmbeddingsBackend(name?: string): IBackendConfig {
        if (name === undefined || name === '') {
            if (this._defaultEmbeddingsBackend === undefined) {
                throw new Error(`Embeddings backend not found: ${name}. No default backend available.`);
            }

            return this._embeddingsBackends[this._defaultEmbeddingsBackend];
        }

        if (this._embeddingsBackends.hasOwnProperty(name)) {
            return this._embeddingsBackends[name];
        }

        if (this._defaultEmbeddingsBackend !== undefined) {
            return this._embeddingsBackends[this._defaultEmbeddingsBackend];
        }

        throw new Error(`Embeddings backend not found: ${name}. No default backend available.`);
    }

    public getAllCompletionBackends(): IBackendConfig[] {
        return Object.values(this._completionBackends);
    }

    public getAllEmbeddingsBackends(): IBackendConfig[] {
        return Object.values(this._embeddingsBackends);
    }

    public removeAllBackends(): this {
        this.removeAllCompletionBackends();
        this.removeAllEmbeddingBackends();
        return this;
    }

    public removeAllCompletionBackends(): this {
        this._completionBackends = {};
        this._defaultCompletionBackend = undefined;
        return this;
    }

    public removeAllEmbeddingBackends(): this {
        this._embeddingsBackends = {};
        this._defaultEmbeddingsBackend = undefined;
        return this;
    }

    public removeCompletionBackend(label: string): this {
        delete this._completionBackends[label];
        if (this._defaultCompletionBackend === label) {
            this._defaultCompletionBackend = Object.keys(this._completionBackends)[0];
        }

        return this;
    }

    public removeEmbeddingsBackend(label: string): this {
        delete this._embeddingsBackends[label];
        if (this._defaultEmbeddingsBackend === label) {
            this._defaultEmbeddingsBackend = Object.keys(this._embeddingsBackends)[0];
        }

        return this;
    }
}

File: python/semantic_kernel/configuration/kernel_config.py
# Copyright (c) Microsoft. All rights reserved.

from typing import Callable, Dict, List, Optional

from semantic_kernel.ai.open_ai.services.azure_open_ai_config import AzureOpenAIConfig
from semantic_kernel.ai.open_ai.services.open_ai_config import OpenAIConfig
from semantic_kernel.configuration.backend_config import BackendConfig
from semantic_kernel.configuration.backend_types import BackendType
from semantic_kernel.diagnostics.verify import Verify
from semantic_kernel.kernel_exception import KernelException
from semantic_kernel.reliability.pass_through_without_retry import (
    PassThroughWithoutRetry,
)
from semantic_kernel.reliability.retry_mechanism import RetryMechanism


class KernelConfig:
    _completion_backends: Dict[str, BackendConfig] = {}
    _embeddings_backends: Dict[str, BackendConfig] = {}
    _default_completion_backend: Optional[str] = None
    _default_embeddings_backend: Optional[str] = None
    _retry_mechanism: RetryMechanism = PassThroughWithoutRetry()

    def add_azure_openai_completion_backend(
        self,
        name: str,
        deployment_name: str,
        endpoint: str,
        api_key: str,
        api_version: str = "2022-12-01",
        overwrite: bool = False,
    ) -> "KernelConfig":
        Verify.not_empty(name, "The backend name is empty")

        if not overwrite and name in self._completion_backends:
            raise KernelException(
                KernelException.ErrorCodes.InvalidBackendConfiguration,
                f"The completion backend cannot be added twice: {name}",
            )

        self._completion_backends[name] = BackendConfig(
            backend_type=BackendType.AzureOpenAI,
            azure_open_ai=AzureOpenAIConfig(
                deployment_name, endpoint, api_key, api_version
            ),
        )

        if self._default_completion_backend is None:
            self._default_completion_backend = name

        return self

    def add_openai_completion_backend(
        self,
        name: str,
        model_id: str,
        api_key: str,
        org_id: Optional[str] = None,
        overwrite: bool = False,
    ) -> "KernelConfig":
        Verify.not_empty(name, "The backend name is empty")

        if not overwrite and name in self._completion_backends:
            raise KernelException(
                KernelException.ErrorCodes.InvalidBackendConfiguration,
                f"The completion backend cannot be added twice: {name}",
            )

        self._completion_backends[name] = BackendConfig(
            backend_type=BackendType.OpenAI,
            open_ai=OpenAIConfig(model_id, api_key, org_id),
        )

        if self._default_completion_backend is None:
            self._default_completion_backend = name

        return self

    def add_azure_open_ai_embeddings_backend(
        self,
        name: str,
        deployment_name: str,
        endpoint: str,
        api_key: str,
        api_version: str = "2022-12-01",
        overwrite: bool = False,
    ) -> "KernelConfig":
        Verify.not_empty(name, "The backend name is empty")

        if not overwrite and name in self._embeddings_backends:
            raise KernelException(
                KernelException.ErrorCodes.InvalidBackendConfiguration,
                f"The embeddings backend cannot be added twice: {name}",
            )

        self._embeddings_backends[name] = BackendConfig(
            backend_type=BackendType.AzureOpenAI,
            azure_open_ai=AzureOpenAIConfig(
                deployment_name, endpoint, api_key, api_version
            ),
        )

        if self._default_embeddings_backend is None:
            self._default_embeddings_backend = name

        return self

    def add_open_ai_embeddings_backend(
        self,
        name: str,
        model_id: str,
        api_key: str,
        org_id: Optional[str] = None,
        overwrite: bool = False,
    ) -> "KernelConfig":
        Verify.not_empty(name, "The backend name is empty")

        if not overwrite and name in self._embeddings_backends:
            raise KernelException(
                KernelException.ErrorCodes.InvalidBackendConfiguration,
                f"The embeddings backend cannot be added twice: {name}",
            )

        self._embeddings_backends[name] = BackendConfig(
            backend_type=BackendType.OpenAI,
            open_ai=OpenAIConfig(model_id, api_key, org_id),
        )

        if self._default_embeddings_backend is None:
            self._default_embeddings_backend = name

        return self

    def has_completion_backend(
        self, name: str, condition: Optional[Callable[[BackendConfig], bool]] = None
    ) -> bool:
        if condition is None:
            return name in self._completion_backends

        return any(
            [name == n and condition(v) for n, v in self._completion_backends.items()]
        )

    def has_embeddings_backend(
        self, name: str, condition: Optional[Callable[[BackendConfig], bool]] = None
    ) -> bool:
        if condition is None:
            return name in self._embeddings_backends

        return any(
            [name == n and condition(v) for n, v in self._embeddings_backends.items()]
        )

    def set_retry_mechanism(
        self, retry_mechanism: Optional[RetryMechanism]
    ) -> "KernelConfig":
        self._retry_mechanism = (
            retry_mechanism
            if retry_mechanism is not None
            else PassThroughWithoutRetry()
        )
        return self

    def set_default_completion_backend(self, name: str) -> "KernelConfig":
        if name not in self._completion_backends:
            raise KernelException(
                KernelException.ErrorCodes.BackendNotFound,
                f"The completions backend doesn't exist: {name}",
            )

        self._default_completion_backend = name
        return self

    @property
    def default_completion_backend(self) -> Optional[str]:
        return self._default_completion_backend

    def set_default_embeddings_backend(self, name: str) -> "KernelConfig":
        if name not in self._embeddings_backends:
            raise KernelException(
                KernelException.ErrorCodes.BackendNotFound,
                f"The embeddings backend doesn't exist: {name}",
            )

        self._default_embeddings_backend = name
        return self

    @property
    def default_embeddings_backend(self) -> Optional[str]:
        return self._default_embeddings_backend

    def get_completion_backend(self, name: Optional[str]) -> BackendConfig:
        if name is None or name.strip() == "":
            if self._default_completion_backend is None:
                raise KernelException(
                    KernelException.ErrorCodes.BackendNotFound,
                    f"Completion backend not found: {name}. "
                    f"No default backend available.",
                )

            return self._completion_backends[self._default_completion_backend]

        if name in self._completion_backends:
            return self._completion_backends[name]

        if self._default_completion_backend is not None:
            return self._completion_backends[self._default_completion_backend]

        raise KernelException(
            KernelException.ErrorCodes.BackendNotFound,
            f"Completion backend not found: {name}. " f"No default backend available.",
        )

    def get_embeddings_backend(self, name: Optional[str]) -> BackendConfig:
        if name is None or name.strip() == "":
            if self._default_embeddings_backend is None:
                raise KernelException(
                    KernelException.ErrorCodes.BackendNotFound,
                    f"Embeddings backend not found: {name}. "
                    f"No default backend available.",
                )

            return self._embeddings_backends[self._default_embeddings_backend]

        if name in self._embeddings_backends:
            return self._embeddings_backends[name]

        if self._default_embeddings_backend is not None:
            return self._embeddings_backends[self._default_embeddings_backend]

        raise KernelException(
            KernelException.ErrorCodes.BackendNotFound,
            f"Embeddings backend not found: {name}. " f"No default backend available.",
        )

    def get_all_completion_backends(self) -> List[BackendConfig]:
        return list(self._completion_backends.values())

    def get_all_embeddings_backends(self) -> List[BackendConfig]:
        return list(self._embeddings_backends.values())

    def remove_completion_backend(self, name: str) -> None:
        if name in self._completion_backends:
            del self._completion_backends[name]

        if name == self._default_completion_backend:
            self._default_completion_backend = (
                list(self._completion_backends.keys())[0]
                if len(self._completion_backends) > 0
                else None
            )

    def remove_embeddings_backend(self, name: str) -> None:
        if name in self._embeddings_backends:
            del self._embeddings_backends[name]

        if name == self._default_embeddings_backend:
            self._default_embeddings_backend = (
                list(self._embeddings_backends.keys())[0]
                if len(self._embeddings_backends) > 0
                else None
            )

    def remove_all_completion_backends(self) -> None:
        self._completion_backends.clear()
        self._default_completion_backend = None

    def remove_all_embeddings_backends(self) -> None:
        self._embeddings_backends.clear()
        self._default_embeddings_backend = None

    def remove_all_backends(self) -> None:
        self.remove_all_completion_backends()
        self.remove_all_embeddings_backends()
